### PR TITLE
sax parser option

### PR DESCRIPTION
**Pull request by vnitinv**
ncclient by default uses DOM parsing. In DOM parsing the whole XML is converted into an lxml object, making it memory intensive. For example for a 100 MB XML file, the lxml object can be around 800 MB.

In the majority of cases, the user only needs to fetch a few tag values from the XML, hence loading full XML using the DOM object is suboptimal. 

On the other hand, SAX parsing will only store the desired tag values. (parsing at run time when XML is being received) 

Wherever we just need to retrieve only a few tag values from XML output, SAX parsing should be preferred over DOM

For 100 MB XML data consuming 800 MB in case of DOM parsing is expected to reduce to 100MB with SAX parser.

In Junos PyEZ example, use_filter is used to decide which parser to be used and filter-xml is input for SAX parsing.

```python
with Device(host='xxxx', user='xxxxx', password='xxxx', use_filter=True) as dev:
    sax_input = "<interface-information><physical-interface><name/></physical-interface></interface- 
    information>"
    dev.rpc.get_interface_information(filter_xml=sax_input)
```

### CODE REVIEW

File: ncclient/transport/parser.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    def callback(self, root, raw):
+        if type(self._session.parser) == DefaultXMLParser:
+            self._session.parser = self._session._device_handler.get_xml_parser(self._session)
+
+    def errback(self, _):
+        pass
+
+
+class SAXFilterXMLNotFoundError(OperationError):
+    def __init__(self, rpc_listner):
```
Comment (einarnn) [inline ncclient/transport/parser.py]: typo: `rpc_listner` -> `rpc_listener`

File: ncclient/transport/parser.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        self._listner = rpc_listner
```
Comment (einarnn) [inline ncclient/transport/parser.py]: typo: `listner` -> `listener`

File: ncclient/transport/parser.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+    def __str__(self):
+        return "SAX filter input xml not provided for listner: %s" % self._listner
```
Comment (einarnn) [inline ncclient/transport/parser.py]: typo: `listner` -> `listener`

File: ncclient/devices/junos.py [hunk 4]
```diff
@@ -110,3 +114,15 @@ def transform_reply(self):
             return reply
         else:
             return reply.encode('UTF-8')
+
+    def get_xml_parser(self, session):
+        # use_filter in device_params can be used to enabled using SAX parsing
+        if self.device_params.get('use_filter', False):
+            l = session.get_listener_instance(SAXParserHandler)
```
Comment (einarnn) [inline ncclient/devices/junos.py]: You've made using the new SAX parser something that only Junos gets without writing another device handler. Any particular reason you didn't integrate into the default device handler? After all, you've refactored the code for all :)

File: ncclient/transport/ssh.py [hunk 5]
```diff
@@ -46,7 +48,6 @@
 BUF_SIZE = 4096
 # v1.0: RFC 4742
 MSG_DELIM = "]]>]]>"
```
Comment (einarnn) [inline ncclient/transport/ssh.py]: should this be left here?

File: ncclient/transport/ssh.py [hunk 6]
```diff
@@ -46,7 +48,6 @@
 BUF_SIZE = 4096
```
Comment (einarnn) [inline ncclient/transport/ssh.py]: should this be left here?

File: ncclient/transport/third_party/junos/parser.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+from ncclient.transport.parser import DefaultXMLParser
+from ncclient.operations import rpc
+from ncclient.transport.parser import SAXFilterXMLNotFoundError
+from ncclient.operations.errors import OperationError
+
+import logging
+logger = logging.getLogger("ncclient.transport.third_party.junos.parser")
+
+# v1.0: RFC 4742
+MSG_DELIM = "]]>]]>"
```
Comment (einarnn) [inline ncclient/transport/third_party/junos/parser.py]: this is common across both parsers; should it be in a separate common module?

File: ncclient/transport/third_party/junos/parser.py [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+MSG_DELIM_LEN = len(MSG_DELIM)
```
Comment (einarnn) [inline ncclient/transport/third_party/junos/parser.py]: this is common across both parsers; should it be in a separate common module?

File: ncclient/transport/third_party/junos/parser.py (hunk 7 above)
Comment (vnitinv) [reply ncclient/transport/third_party/junos/parser.py]: Fixed via https://github.com/ncclient/ncclient/pull/324/commits/9fcad710757bf774bd672d8253e1c42b6083eda1

File: ncclient/transport/parser.py (hunk 3 above)
Comment (vnitinv) [reply ncclient/transport/parser.py]: Fixed via https://github.com/ncclient/ncclient/pull/324/commits/9fcad710757bf774bd672d8253e1c42b6083eda1

File: ncclient/transport/parser.py (hunk 2 above)
Comment (vnitinv) [reply ncclient/transport/parser.py]: Fixed via https://github.com/ncclient/ncclient/pull/324/commits/9fcad710757bf774bd672d8253e1c42b6083eda1

File: ncclient/transport/parser.py (hunk 1 above)
Comment (vnitinv) [reply ncclient/transport/parser.py]: Fixed via https://github.com/ncclient/ncclient/pull/324/commits/9fcad710757bf774bd672d8253e1c42b6083eda1

File: ncclient/transport/ssh.py (hunk 5 above)
Comment (vnitinv) [reply ncclient/transport/ssh.py]: Yes, this is commonplace. I find it good to be here.

File: ncclient/transport/ssh.py (hunk 6 above)
Comment (vnitinv) [reply ncclient/transport/ssh.py]: Yes, this is commonplace. I find it good to be here.

**Comment by vnitinv** [comment]
@einarnn I have addressed your comment. Please do let me know if any other changes are required.

**Pull request merged**